### ISSUE TITLE
Support time recording for Rally Cross events

### ISSUE THREAD

**Issue by maxbechtold** [soong-construction/dirt-rally-time-recorder#19]
I'm unsure if this is interesting to anyone, but I could add support for Rally Cross to *dirt-rally-time-recorder*.  

There's no stage time in RX, but rather lap time, but this is also reported via UDP if memory serves. The tool could record any completed lap, or perhaps only the fastest in a (finished) session/race.

**Comment by gotleg** [comment]
This is exactly what I'm looking for 👍 
I was initially looking for an api maybe over racenet to get this info but it doesn't look possible.
Using UDP data could be an answer

**Comment by maxbechtold** [comment]
Good to hear! Do you mean you will try this tool for stage time recording? Or are you rooting for adding Rally cross support? There has been no effort in that direction as of now.

Indeed I don't think there's a way to extract laptimes from the game (which seems to treat best times per class rather than per car).

**Comment by gotleg** [comment]
Hi @maxbechtold ,

Actuely i Only play Rallycross and I'm interested in tracking races Times especially racing with Friends yo comparé Times ans track out progression. 

Thanks to your project I managed to code a small nodejs program that listens to UDP Packets, and send each lap/lap_time into a DB.
Each pilot can install this program to report times in a centralized database. Then I'll generate dashboard based on lap_times.

Now that I found a correct way to track lap_times I Can easily add support for Rallycross on dirt-rally-time-recorder.
I'm not very familiar with python but I'll try to find some time to add it.

**Comment by maxbechtold** [comment]
Sounds like a project I'd like to follow, is it OSS? Then maybe this issue is resolved by placing a link to your project.

Otherwise let's start a discussion here on how RX could be included in *dirt-rally-time-recorder*, from the top of my head:

1. record any completed lap, or only the fastest in a (finished) session/race?
2. treat race modes differently (single practice vs. heats & (semi) finals)?
3. treat lap 1 differently (which can start from an off-track grid)?
4. treat Joker laps differently (mostly slower than regular laps)?
5. is support for both DiRT Rally 2.0 as well as the original version possible?
6. list/export RX laptimes separate from stage times (since these are unrelated)?

Since a bunch of new tracks and cars must be identified, I could assist with providing the telemetry data for those.

**Comment by gotleg** [comment]
Hi, 

Yes my app will be open source but for now it is dirtt as pas just exploring 😜
I'll clean this UP and publish it probably tonight or tomorow.

1- I record every last_lap_time with trackname (based on track_length) and lap number + timestamp.
2- It would be great but I have no Idea how to detect this. Maybe WE Can already track total race Time and lap count to regroupe.
3 - yes possible based on lap number
4- yes Maybe based on lap_distance.
5- I Only tried DR2 but Packets seems close this shoul be possible
6- based on track length i Can already identify wich track we are driving

**Comment by gotleg** [comment]
By the way one part of my project is providing an API to report personnal laps and then create WEB dashboard to analyses performances on tracks, during time, ...
I'll probably also add "community features" like create a team with other pilots and provide a team dashboard.
Maybe I can focus my energy on this API + dashboards and help you add rally cross support + API calls so we can have complementary features :)

**Comment by maxbechtold** [comment]
> 2- It would be great but I have no Idea how to detect this. Maybe WE Can already track total race Time and lap count to regroup.

Yeah, I don't see how this can be pulled from the telemetry with certainty. But maybe it's not needed anyway.

> 6- based on track length i Can already identify wich track we are driving

Of course the tracks can be identified by length (like for stage rallies), I meant that there might be a script like `list-rx-laptimes.bat` which only shows RX laps. RX times should also be stored in a table separate from stage times, e.g. top speed doesn't seem to make sense per lap.

> By the way one part of my project is providing an API to report personnal laps and then create WEB dashboard to analyses performances on tracks, during time, ...

Interesting, but an entirely different topic. This tool started out as a web solution (which was broken since the server was shut down). Considering this could be seen as poaching in Codemaster's territory (Racenet), I don't think *dirt-rally-time-recorder* should go that way. GDPR and data privacy in general is also something that would have to be taken into account. 

Instead, I've been thinking about a peer-to-peer solution where users can exchange their best times without a central server, just by exchanging db files. I think it's less effort than going client/server.
